sematic/caching/caching.py

```diff
@@ -0,0 +1,106 @@
+# Standard Library
+import hashlib
+import json
+import logging
+from typing import Any, Callable, Optional, Union
+
+# Sematic
+from sematic.abstract_future import AbstractFuture
+from sematic.types.serialization import value_to_json_encodable
+
+CacheNamespaceCallable = Callable[[AbstractFuture], str]
+CacheNamespace = Optional[Union[str, CacheNamespaceCallable]]
+
+logger = logging.getLogger(__name__)
+
+
+def get_future_cache_key(
+    cache_namespace: CacheNamespace, future: AbstractFuture
+) -> str:
+    """
+    Generates a cache key that can be used to uniquely identify the output value of a
+    deterministic function.
+
+    Parameters
+    ----------
+    cache_namespace: CacheNamespace
+        A string or a `Callable` which returns a string, which is used as the cache key
+        namespace.
+    future: AbstractFuture
+        The future for which to calculate the cache key.
+
+    Returns
+    -------
+    A cache key string that can be used to uniquely identify the output value of a
+    deterministic function.
+    """
+    # this may raise, so do it early
+    cache_namespace = _resolve_namespace(cache_namespace=cache_namespace, future=future)
```

Comment:
I think the namespace should be shared across a whole resolution. So, if it is only ever called with the root future, that's fine (just document it somewhere). But this code looks as if it is intended to be called per-future. If that's the case, I think we should change that.

Reasoning:
It is conceptually clear to say something to the effect of:
- the namespace is shared across a whole resolution
- each future gets a hash specific to that future, which is nested under the shared namespace
- the final cache key is a combination of the namespace and the future-specific hash

There are more blurry lines if you instead have this:
- the namespace is sometimes a constant for the whole resolution, but might also vary per future
- each future gets a future-specific hash
- the final cache key is a combination of the namespace and the future-specific hash

If each future is already getting a future-specific hash, why add any other future-specific logic for the namespace? How should we determine when some future-specific piece gets put into the future-specific namespace calculation or just the pure future-specific hash?

Reply:
It's possible that in order to get the `git` sha, you need access to a func. In that case, you can just call your namespace generating function on the root future and pass the resulting namspace string to be used for the rest of the resolution